File: c/chapitre_4/exo41.c
/*
Écrire un programme qui crée deux processus. Le processus père
ouvre un ﬁchier texte en lecture. On suppose que le ﬁchier est composé de mots formés de
caractères alphabétiques séparés par des espaces. Le processus ﬁls saisit un mot au clavier. Le
processus père recherche le mot dans le ﬁchier, et transmet au ﬁls la valeur 1 si le mot est dans
le ﬁchier, et 0 sinon. Le ﬁls aﬃche le résultat.
*/

#include <stdio.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <unistd.h>
#include <string.h>

#define BUFF 256

int main(int argc , char **argv)
{
    int resultat , trouver = 0 ;
    int tubeMot[2];
    int tubeRep[2] ;
    char tabFils[BUFF] , tabPere[BUFF] ;
    FILE *fichier ;

    // Création des 2 pipe
    pipe(tubeMot);
    pipe(tubeRep);

    if (argc != 2 )
    {
        printf("Faut preciser le nom du fichier\n");
        exit(-1);
    }
    

    __pid_t pid = fork();
    
    if( pid == -1 )
    {
        perror("Erreur Fork\n");
        exit(-1);
    }
    else if (pid == 0 )
    {
        close(tubeMot[0]);
        close(tubeRep[1]);

        printf("Entrez le mot : ");
        scanf("%s",tabFils) ;

        // Envoie du message au Père
        write(tubeMot[1] , tabFils , BUFF ) ;

        // Verification du resultat venant du père
        read(tubeRep[0] , &resultat , sizeof(int)) ;
        if (resultat)
        {
            printf("Mot trouvé\n");
        }
        else
        {
            printf("Mot pas encore touvé\n") ;
        }
        
        exit(0);
    }
    else
    {
        fichier = fopen(argv[1] , "r" ) ;
        char ligne[500] ;
        close(tubeMot[1]);
        close(tubeRep[0]) ;

    // Reception du message venant du fils
        read(tubeMot[0] , tabPere , BUFF ) ;
    // recherche le mot dans le fichier et return 1 au fils si trouvé et 0 si non
        while ( fscanf(fichier , "%s" , ligne ) != EOF )
        {
            if (strcmp(ligne , tabPere ) == 0 )
            {
                trouver = 1 ;
                break;
            }
        }

        fclose(fichier) ;

        // Envoie du reponse au fils
        write(tubeRep[1] , &trouver , sizeof(int) ) ;

        wait(NULL);
    }

    
    return (0) ;
}

File: python/chapitre_4/exo41.py
"""
Écrire un programme qui crée deux processus. Le processus père
ouvre un ﬁchier texte en lecture. On suppose que le ﬁchier est composé de mots formés de
caractères alphabétiques séparés par des espaces. Le processus ﬁls saisit un mot au clavier. Le
processus père recherche le mot dans le ﬁchier, et transmet au ﬁls la valeur 1 si le mot est dans
le ﬁchier, et 0 sinon. Le ﬁls aﬃche le résultat.
"""

import os
import sys
import struct

BUFF = 256

def main():
    resultat = 0
    trouver = 0

    # Création des 2 pipe
    tube_mot_r, tube_mot_w = os.pipe()
    tube_rep_r, tube_rep_w = os.pipe()

    if len(sys.argv) != 2:
        print("Faut preciser le nom du fichier")
        sys.exit(-1)

    pid = os.fork()

    if pid == -1:
        print("Erreur Fork")
        sys.exit(-1)
    elif pid == 0:
        os.close(tube_mot_r)
        os.close(tube_rep_w)

        tab_fils = input("Entrez le mot : ")

        # Envoie du message au Père
        os.write(tube_mot_w, tab_fils.encode().ljust(BUFF, b'\x00'))

        # Verification du resultat venant du père
        data = os.read(tube_rep_r, 4)
        resultat = struct.unpack('i', data)[0]
        if resultat:
            print("Mot trouvé")
        else:
            print("Mot pas encore trouvé")

        sys.exit(0)
    else:
        os.close(tube_mot_w)
        os.close(tube_rep_r)

        # Reception du message venant du fils
        tab_pere = os.read(tube_mot_r, BUFF).decode().rstrip('\x00').strip()

        # recherche le mot dans le fichier et return 1 au fils si trouvé et 0 si non
        with open(sys.argv[1], 'r') as fichier:
            for ligne in fichier.read().split():
                if ligne == tab_pere:
                    trouver = 1
                    break

        # Envoie du reponse au fils
        os.write(tube_rep_w, struct.pack('i', trouver))

        os.wait()

    return 0

main()